### PR TITLE
Compile time generated type info for bundles

### PR DESCRIPTION
**Pull request by YohDeadfall**
The change alters how static type data for bundles are generated, so no allocations and sorting should happen any more during an app execution. It requires some nightly features for const evaluation one of which I don't like. It's `transmute` used to compare type identifiers since ordering isn't defined as a constant function. Others should be settled soon in `stable` toolchain, so will be removed at that point.

Another thing which bothers me is consants declared by `Bundle` tuple, but decided to use them instead of local static items in functions since the last is just another layer of abstraction and found that that's how some things done in Rust's standard library.

Unfortunately, I don't know how to run old tests from `bench.rs` and will highly appreciate any information on it. My attemps to make them work or even be detected ended by nothing.

### CODE REVIEW

File: crates/bevy_ecs/src/core/entity_builder.rs [hunk 1]
```diff
@@ -121,7 +121,10 @@ impl EntityBuilder {
             } else {
                 max_align as *mut _
             };
-            for (ty, offset) in self.info.drain(..) {
+            for ty in self.info.drain(..) {
+                let offset = *self.offsets
+                    .get(&ty.id())
+                    .unwrap();
```
Comment (YohDeadfall) [inline crates/bevy_ecs/src/core/entity_builder.rs:127]: Perfectly, this should be optimized too, but for now I have no idea how to simplify it. Will think a bit later on it.

**Comment by alice-i-cecile** [comment]
@cart is working on an ECS rewrite at the moment that should have its own PR any day now. Probably worth moving this to draft for now, and then seeing what bits still apply once that code is visible <3

**Comment by YohDeadfall** [comment]
Good to know it. We had a conversation a while ago privately during which he mentioned on what I can work, and ECS is mine interest here. Had a few more ideas on memory optimizations, but let's wait for @cart to publish his work first. If there's anything left with which I could help, let me know.

> Probably worth moving this to draft for now

Sadly, there's no option to make something a draft when it was published.

**Comment by cart** [comment]
For future reference:

image

**Comment by YohDeadfall** [comment]
Expected to see that option in any other place except that, thanks!

**Comment by cart** [comment]
Yeah we can't use nightly features in bevy, but im still very curious to see how this all works. Bundles have been simplified a lot in the rewrite, but there might still be aspects that from this pr that apply there. That pr should be ready in the next couple of days.

**Comment by cart** [comment]
I'm going to close this for now. Its definitely interesting, but #1525 largely solves the "computational cost" of TypeInfo. Making it fully const would be nice, but thats likely off the table on stable for the foreseeable future.

**Comment by YohDeadfall** [comment]
Sure, but I still would like to bring that feature to Bevy. Should I open a new PR when things stabilize or continue this one after rebase?

**Pull request closed without merging**